### PR TITLE
Improve verbose VFS logging

### LINKED ISSUE #14361: Improve verbose VFS logging
- [x] Log number of watched hierarchies
- [x] Reinstate the build operation based test in `ChangesDuringTheBuildFileSystemWatchingIntegrationTest` (see https://github.com/gradle/gradle/pull/14303/files/7d2a5259404c79d9750d20ac7ddf45e548ab7a37#diff-50647a4a1268ccee03a38b6de88390e3)
- [x] Update `FileSystemWatchingSoakTest` to assert the number of invalidated files again (see https://github.com/gradle/gradle/pull/14303#pullrequestreview-478542291)
- [x] Mention verbose logging flag in daemon documentation

---
cc: @gradle/execution

### PR DESCRIPTION
**Pull request by lptr**
Fixes #14361

### CODE REVIEW

**Comment by lptr** [comment]
@bot-gradle test QFL please.

**Comment by lptr** [comment]
@bot-gradle test QF please.

File: subprojects/docs/src/docs/userguide/running-builds/gradle_daemon.adoc [hunk 1]
```diff
@@ -232,31 +232,51 @@ This enables watching the file system for all builds, unless explicitly disabled
...
+> Task :processResources NO-SOURCE
+> Task :classes UP-TO-DATE
+> Task :jar UP-TO-DATE
+> Task :assemble UP-TO-DATE
+
+BUILD SUCCESSFUL in 58ms
+1 actionable task: 1 up-to-date
+Received 5 file system events during the current build while watching 1 hierarchies
+Virtual file system retains information about 3 files, 2 directories and 2 missing files until next build
+----
```
Comment (lptr) [inline subprojects/docs/src/docs/userguide/running-builds/gradle_daemon.adoc:258]: Ideally this should come from actually executing a build, but I don't see how we could capture reproducible numbers here.

**Comment by lptr** [comment]
@bot-gradle test QF please.

**Comment by lptr** [comment]
@bot-gradle test QF please.

**Comment by lptr** [comment]
@bot-gradle test this please.

**Comment by lptr** [comment]
@bot-gradle test this please.

File: subprojects/docs/src/docs/userguide/running-builds/gradle_daemon.adoc [hunk 2]
```diff
@@ -232,31 +232,51 @@ This enables watching the file system for all builds, unless explicitly disabled
 === Troubleshooting file system watching
 
 Limitations::
-We are working on removing the following limitations to make file system watching production ready.
+File system watching currently has the following limitations:
 - If you have symlinks in your build, you won’t get the performance benefits for those locations (we plan to change this in the future).
```
Comment (wolfs) [inline subprojects/docs/src/docs/userguide/running-builds/gradle_daemon.adoc:236]:
```suggestion
- If you have symlinks in your build, you won’t get the performance benefits for those locations.
```

I am not sure we are ever going to tackle this.

File: subprojects/file-watching/src/integTest/groovy/org/gradle/internal/watch/ChangesDuringTheBuildFileSystemWatchingIntegrationTest.groovy [hunk 3]
```diff
@@ -102,47 +136,91 @@ class ChangesDuringTheBuildFileSystemWatchingIntegrationTest extends AbstractFil
...
         outputFile.text == "changedAgain"
-        vfsLogs.receivedFileSystemEventsSinceLastBuild >= 0
-        vfsLogs.receivedFileSystemEventsInCurrentBuild >= 0
-        vfsLogs.retainedFilesSinceLastBuild > 0
-        vfsLogs.retainedFilesInCurrentBuild > 0
+        projectFilesInVfs == 2
     }
 
-    private void runWithFileSystemWatchingAndMakeChangesWhen(String task, String expectedCall, Closure action) {
+    private void runWithRetentionAndDoChangesWhen(String task, String expectedCall, Closure action) {
```
Comment (wolfs) [inline subprojects/file-watching/src/integTest/groovy/org/gradle/internal/watch/ChangesDuringTheBuildFileSystemWatchingIntegrationTest.groovy:168]:
```suggestion
    private void runWithFileSystemWatchingAndMakeChangesWhen(String task, String expectedCall, Closure action) {
```

Let's keep the new method name.

File: subprojects/file-watching/src/integTest/groovy/org/gradle/internal/watch/ChangesDuringTheBuildFileSystemWatchingIntegrationTest.groovy [hunk 4]
```diff
@@ -102,47 +136,91 @@ class ChangesDuringTheBuildFileSystemWatchingIntegrationTest extends AbstractFil
...
     }
+
+    int getProjectFilesInVfs() {
+        def retainedInformation = result.getOutputLineThatContains("Project files in VFS: ")
+        def numberMatcher = retainedInformation =~ /Project files in VFS: (\d+)/
+        return numberMatcher[0][1] as int
+
+    }
+
+    void setupFileEventsLogging() {
```
Comment (wolfs) [inline subprojects/file-watching/src/integTest/groovy/org/gradle/internal/watch/ChangesDuringTheBuildFileSystemWatchingIntegrationTest.groovy:184]: We don't need the file events logging, do we? This should be handled by the verbose logging already.

File: subprojects/docs/src/docs/userguide/running-builds/gradle_daemon.adoc [hunk 5]
```diff
@@ -232,31 +232,51 @@ This enables watching the file system for all builds, unless explicitly disabled
...
 - When multiple daemons are running, the idle ones can pick up the changes produced by the others and create large log files with lots of debug info about the changes.
 - On Windows, we don’t support SUBST and network drives (they might work, but we don’t test them yet).
 
-Gradle does not pick up some of my changes.::
+Enable verbose logging::
+You can instruct Gradle to some more information about the state of the virtual file system, and the events received from the file system using the `org.gradle.vfs.verbose` flag.
```
Comment (wolfs) [inline subprojects/docs/src/docs/userguide/running-builds/gradle_daemon.adoc:241]: The flag should also go into `build_environment.adoc`.

**Review by wolfs — CHANGES_REQUESTED** [review]

File: subprojects/file-watching/src/integTest/groovy/org/gradle/internal/watch/ChangesDuringTheBuildFileSystemWatchingIntegrationTest.groovy (hunk 3 above)
Comment (lptr) [reply subprojects/file-watching/src/integTest/groovy/org/gradle/internal/watch/ChangesDuringTheBuildFileSystemWatchingIntegrationTest.groovy:168]: Right.

File: subprojects/file-watching/src/integTest/groovy/org/gradle/internal/watch/ChangesDuringTheBuildFileSystemWatchingIntegrationTest.groovy (hunk 4 above)
Comment (lptr) [reply subprojects/file-watching/src/integTest/groovy/org/gradle/internal/watch/ChangesDuringTheBuildFileSystemWatchingIntegrationTest.groovy]: Removed.

File: subprojects/docs/src/docs/userguide/running-builds/gradle_daemon.adoc (hunk 5 above)
Comment (lptr) [reply subprojects/docs/src/docs/userguide/running-builds/gradle_daemon.adoc:241]: Added. This list should be generated, though.

**Review by wolfs — APPROVED** [review]
LGTM! Can you run the soak tests on the three OSes to make sure they pass?

**Comment by lptr** [comment]
Soak tests are passing:

- Windows: https://builds.gradle.org/buildConfiguration/Gradle_Check_Soak_9_0/37720661
- Linux: https://builds.gradle.org/buildConfiguration/Gradle_Check_Soak_8_0/37720267
- macOS: https://builds.gradle.org/buildConfiguration/Gradle_Check_Soak_35_0/37720213

**Pull request merged**